gitgud/levels/util.py

```diff
@@ -194,15 +192,67 @@ def test_level(level, test):
 
     return True
 
+class NamedList:
+    # names is a list populated with type str, items is a list populated with any type 
+    def __init__(self, names, items):
+        assert len(names) == len(items)
+        self._name_dict = {name:index for index, name in enumerate(names)}
+        self._items = items
+    
+    def __getitem__(self, query):
+        if isinstance(query, int):
+            return self._items[query]
+        elif isinstance(query, str):
+            return self._items[self._name_dict[query]]
+        else:
+            raise ValueError('Bad key type.')
+
+    def __iter__(self):
+        return self._items.__iter__()
+
+    def __len__(self):
+        return len(self._items)
+
+    def __setitem__(self, key, item):
+        if isinstance(key, str):
+            self._name_dict[key] = len(self._items)
+            self._items.append(item)
+        else:
+            raise TypeError
+
+    def __contains__(self, key):
+        if isinstance(key, str):
+            return key in self._name_dict.keys()
+        else:
+            return key in self._items()
 
-class Level:
+    def values(self):
+        return self._items
+    
+    def keys(self):
+        return self._name_dict.keys()
+
+class Level(NamedList):
     def __init__(self, name, challenges):
         self.name = name
-        self.challenges = OrderedDict()
-        for challenge in challenges.values():
+        self._name_dict = {challenge.name:index for index, challenge in enumerate(challenges)}
+        self._items = challenges
+
+        for challenge in challenges:
             challenge.level = self
-            self.challenges[challenge.name] = challenge
 
+class AllLevels(NamedList):
+    def __init__(self, levels):
+        self._name_dict = {level.name:index for index, level in enumerate(levels)}
+        self._items = levels
+    
+    def _add_next_challenges(self):
```

Comment:
You can remove this method definition entirely and just put the code in the constructor since we'll want it to be called every time no matter what